modules/lang-painless/src/main/java/org/elasticsearch/painless/api/Augmentation.java

```diff
@@ -552,4 +553,78 @@ public static String decodeBase64(String receiver) {
         // O(N) or faster depending on implementation
         return result.toArray(new String[0]);
     }
+
+    /**
+     * Access values in nested containers with a dot separated path.  Path elements are treated
+     * as strings for Maps and integers for Lists.  Throws 'IllegalArgumentException' if path does
+     * not exist. Throws 'NumberFormatException' if an path element for a List is not an integer.
+     */
+    private static Object getByPath(Object receiver, String path) {
+        Object value = getByPath(receiver, path, new missingValue());
+        if (value instanceof missingValue) {
+            throw new IllegalArgumentException("Could not find value at path [" + path + "]");
+        }
+        return value;
+    }
+
+    /**
+     * Same as 'getByPath(object, String)' but returns defaultValue if path does not exist.
+     */
+    private static Object getByPath(Object receiver, String path, Object defaultValue) {
+        Object next = receiver;
+        String[] elements = path.split("\\.");
```

Comment:
While an iterative solution does work, I think we can eliminate the double double casting necessary in the first call (needing to check instanceof on the receiver even though we just knew the type in the outer call).

Instead of splitting the entire path, I think there should be 2 internal methods, which take Map or List as the receiver, and have an additional argument, the current index in the path. You can then use `indexOf` along with that current index to find the location of the next dot, and extract the current path element.

So there would be a few methods:
* one taking Map, this would have the default calling logic based on whether the map contains the next element, or there is no next path element
* one taking List, this would have the default calling logic based on length of the list, or there is no next path element
* an internal dispatch method that would take Object, which the other two methods would call when they successfully extract the next element

Reply:
Changed to recursion as requested.